### ISSUE TITLE
How to avoid the grad from labels influencing the info_modules?

### ISSUE THREAD

**Issue by godissaw** [Philip-Bachman/amdim-public#2]
In your project, the cross_entropy is used to train the classifier,
`loss_cls = (loss_xent(lgt_glb_mlp, labels) + loss_xent(lgt_glb_lin, labels))`
and the losses are accumulated directly.
`loss_opt = loss_inf + loss_cls`

How could you avoid the probability that the labels are leaked to the unsupervised learning modules?

**Comment by Philip-Bachman** [comment]
All features passed into the classifiers are detached from the computation graph before running through the classifiers. This prevents gradient from flowing back into the encoder through those features. See the code: [`here`](https://github.com/Philip-Bachman/amdim-public/blob/master/model.py#L162)

**Comment by godissaw** [comment]
> All features passed into the classifiers are detached from the computation graph before running through the classifiers. This prevents gradient from flowing back into the encoder through those features. See the code: [`here`](https://github.com/Philip-Bachman/amdim-public/blob/master/model.py#L162)
> 
> We have also trained the model with no online classifier monitoring to verify that the labels are not affecting the encoder. We provide the code with online classifier monitoring since it is more convenient for rapid iteration. If you modify the code so that features passed to the classifiers are not detached from the computation graph, you will see that behaviour and performance of the model changes significantly.

Thanks a lot, your answer solved my problem entierly.
